File: modules/agent/src/serializers/individual.py
"""Serializers for individual-related models."""

from serializers.common import contact_to_dict


def get_industries(ind) -> list:
    """Extract industry names from individual's account."""
    if not ind.account or not ind.account.industries:
        return []
    return [industry.name for industry in ind.account.industries]


def get_projects(ind) -> list:
    """Extract projects info from individual's account."""
    if not ind.account or not ind.account.projects:
        return []
    return [{"id": p.id, "name": p.name} for p in ind.account.projects]


def get_related_orgs_from_contacts(contacts) -> list:
    """Extract unique organizations from contacts."""
    seen_ids = set()
    orgs = []
    for contact in contacts:
        for org in (contact.organizations or []):
            if org.id not in seen_ids:
                seen_ids.add(org.id)
                orgs.append({"id": org.id, "name": org.name, "type": "organization"})
    return orgs


def ind_to_list_response(ind) -> dict:
    """Convert individual ORM to response dictionary - optimized for list/table view."""
    return {
        "id": ind.id,
        "first_name": ind.first_name,
        "last_name": ind.last_name,
        "title": ind.title,
        "industries": get_industries(ind),
        "email": ind.email,
        "phone": ind.phone,
        "updated_at": ind.updated_at.isoformat() if ind.updated_at else None,
    }


def ind_to_search_response(ind) -> dict:
    """Convert individual ORM to response dictionary - minimal for search/autocomplete."""
    return {
        "id": ind.id,
        "first_name": ind.first_name,
        "last_name": ind.last_name,
        "email": ind.email,
    }


def ind_to_detail_response(ind, contacts=None, include_research=False) -> dict:
    """Convert individual ORM to full response dictionary - for detail/edit views."""
    result = {
        "id": ind.id,
        "first_name": ind.first_name,
        "last_name": ind.last_name,
        "email": ind.email,
        "phone": ind.phone,
        "linkedin_url": ind.linkedin_url,
        "title": ind.title,
        "description": ind.description,
        "industries": get_industries(ind),
        "projects": get_projects(ind),
        "twitter_url": ind.twitter_url,
        "github_url": ind.github_url,
        "bio": ind.bio,
        "seniority_level": ind.seniority_level,
        "department": ind.department,
        "is_decision_maker": ind.is_decision_maker,
        "reports_to_id": ind.reports_to_id,
        "created_at": ind.created_at.isoformat() if ind.created_at else None,
        "updated_at": ind.updated_at.isoformat() if ind.updated_at else None,
    }

    if contacts is not None:
        result["contacts"] = [contact_to_dict(c) for c in contacts]
        result["relationships"] = get_related_orgs_from_contacts(contacts)

    if include_research:
        if ind.reports_to:
            result["reports_to"] = {
                "id": ind.reports_to.id,
                "first_name": ind.reports_to.first_name,
                "last_name": ind.reports_to.last_name,
            }
        result["direct_reports"] = [
            {"id": dr.id, "first_name": dr.first_name, "last_name": dr.last_name, "title": dr.title}
            for dr in (ind.direct_reports or [])
        ]

    return result

File: modules/agent-go/internal/serializers/individual.go
package serializers

import (
	"time"

	"github.com/pina-colada-co/agent-go/internal/models"
)

// IndividualListResponse represents an individual in list view
type IndividualListResponse struct {
	ID             int64           `json:"id"`
	FirstName      string          `json:"first_name"`
	LastName       string          `json:"last_name"`
	Email          *string         `json:"email"`
	Phone          *string         `json:"phone"`
	Title          *string         `json:"title"`
	SeniorityLevel *string         `json:"seniority_level"`
	Department     *string         `json:"department"`
	IsDecisionMaker *bool          `json:"is_decision_maker"`
	Industries     []IndustryBrief `json:"industries"`
	CreatedAt      time.Time       `json:"created_at"`
	UpdatedAt      time.Time       `json:"updated_at"`
}

// IndividualDetailResponse represents an individual in detail view
type IndividualDetailResponse struct {
	IndividualListResponse
	LinkedInURL *string        `json:"linkedin_url"`
	TwitterURL  *string        `json:"twitter_url"`
	GithubURL   *string        `json:"github_url"`
	Bio         *string        `json:"bio"`
	Description *string        `json:"description"`
	Contacts    []ContactBrief `json:"contacts"`
	Projects    []ProjectBrief `json:"projects"`
}

// IndividualToListResponse converts Individual model to list response
func IndividualToListResponse(ind *models.Individual) IndividualListResponse {
	resp := IndividualListResponse{
		ID:             ind.ID,
		FirstName:      ind.FirstName,
		LastName:       ind.LastName,
		Email:          ind.Email,
		Phone:          ind.Phone,
		Title:          ind.Title,
		SeniorityLevel: ind.SeniorityLevel,
		Department:     ind.Department,
		IsDecisionMaker: ind.IsDecisionMaker,
		CreatedAt:      ind.CreatedAt,
		UpdatedAt:      ind.UpdatedAt,
	}

	if ind.Industries != nil {
		resp.Industries = make([]IndustryBrief, len(ind.Industries))
		for i, industry := range ind.Industries {
			resp.Industries[i] = IndustryBrief{ID: industry.ID, Name: industry.Name}
		}
	}

	return resp
}

// IndividualToDetailResponse converts Individual model to detail response
func IndividualToDetailResponse(ind *models.Individual) IndividualDetailResponse {
	resp := IndividualDetailResponse{
		IndividualListResponse: IndividualToListResponse(ind),
		LinkedInURL:           ind.LinkedInURL,
		TwitterURL:            ind.TwitterURL,
		GithubURL:             ind.GithubURL,
		Bio:                   ind.Bio,
		Description:           ind.Description,
	}

	if ind.Contacts != nil {
		resp.Contacts = make([]ContactBrief, len(ind.Contacts))
		for i, c := range ind.Contacts {
			resp.Contacts[i] = ContactBrief{
				ID:        c.ID,
				FirstName: c.FirstName,
				LastName:  c.LastName,
				Email:     c.Email,
				Phone:     c.Phone,
				Title:     c.Title,
				IsPrimary: c.IsPrimary,
			}
		}
	}

	if ind.Projects != nil {
		resp.Projects = make([]ProjectBrief, len(ind.Projects))
		for i, p := range ind.Projects {
			resp.Projects[i] = ProjectBrief{ID: p.ID, Name: p.Name}
		}
	}

	return resp
}
